### PR TITLE
Refactor:

### PR DESCRIPTION
**Pull request by mohsen2014**
- Add typescript
- Apply some refactor on app.js

### CODE REVIEW

**Comment by ShrimpCryptid** [comment]
Thanks for the pull request!
Any particular reason for having the new files be typescript rather than javascript?

**Comment by mohsen2014** [comment]
Yes, I have found from experience that working with ،Typescript is much better and easier over time and eliminates many logical errors.

**Comment by mohsen2014** [comment]
I want to work on this project, but first I have to refactor it to understand your code. Please accept my pull request

**Comment by ShrimpCryptid** [comment]
Yes, of course, thanks for following up! The change looks good but testing has been on my back burner for a few days. I'll try and schedule time to test it tonight.

**Comment by ShrimpCryptid** [comment]
Firstly, thank you again for your patience and for your willingness to look through this project! It really means a lot to me that you want to support it. ☺

When I run the change locally, I get the following error message:

```
TypeError: Cannot read properties of undefined (reading 'joinLobby')

.../secret-hitler-online/secret-hitler-online-interface/src/App.js:441
shouldJoinButtonBeEnabled
  431 |    };
  432 | 
  433 |    shouldJoinButtonBeEnabled() {
> 434 |        return (this.state.joinLobby.length === LOBBY_CODE_LENGTH) && (this.state.joinName.length !== 0)
  435 |    }
  436 | 
  437 |    shouldCreateLobbyButtonBeEnabled() {
  ```
  
The `shouldJoinButtonBeEnabled` method cannot access `state`, and several other methods have this same issue. It looks the new context in which these methods are being called (from `JoinLobby`) is causing failures when accessing `state`, as they're not bound to App.js. 

I've previously solved this by binding the methods to App.js in the constructor using `this.methodName= this.methodName.bind(this);` A similar solution can be found in [this StackOverflow thread](https://stackoverflow.com/questions/35287949/react-with-es7-uncaught-typeerror-cannot-read-property-state-of-undefined). Feel free to suggest an alternative if you find another solution.
 
I also wanted to ask if you were able to run this locally and test these changes? If not, I can provide some instructions on how to set up the project for testing. (This is definitely something I need to include in a README!)

File: secret-hitler-online-interface/src/App.js [hunk 1]
```diff
@@ -431,11 +431,12 @@ class App extends Component {
...
 
     shouldJoinButtonBeEnabled() {
-        return (this.state.joinLobby.length === LOBBY_CODE_LENGTH) && (this.state.joinName.length !== 0)
+        return this?.state?.joinLobby?.length === LOBBY_CODE_LENGTH && 
+            this?.state?.joinName?.length !== 0;
     }
 
     shouldCreateLobbyButtonBeEnabled() {
-        return (this.state.createLobbyName.length !== 0);
+        return this?.state?.createLobbyName.length !== 0;
```
Comment (ShrimpCryptid) [inline secret-hitler-online-interface/src/App.js:439]: This line currently results in the Create Lobby button always being enabled.
Changing it back to `return this.state.createLobbyName.length !== 0` and adding the line `this.shouldCreateLobbyButtonBeEnabled = this.shouldCreateLobbyButtonBeEnabled.bind(this);` in the constructor appears to fix this issue.

File: secret-hitler-online-interface/src/App.js [hunk 2]
```diff
@@ -431,11 +431,12 @@ class App extends Component {
     };
 
     shouldJoinButtonBeEnabled() {
-        return (this.state.joinLobby.length === LOBBY_CODE_LENGTH) && (this.state.joinName.length !== 0)
```
Comment (ShrimpCryptid) [inline secret-hitler-online-interface/src/App.js:482]: See the similar note below. Currently, the Join button is always disabled, even when putting in a name and lobby code.

File: secret-hitler-online-interface/src/App.js [hunk 3]
```diff
@@ -7,7 +7,7 @@ import MaxLengthTextField from "./util/MaxLengthTextField";
 import CustomAlert from "./custom-alert/CustomAlert";
 import RoleAlert from "./custom-alert/RoleAlert";
 import EventBar from "./event-bar/EventBar";
```
Comment (ShrimpCryptid) [inline secret-hitler-online-interface/src/App.js:9]: May want to clean up unused imports after moving them into gamePage.ts

**Review by ShrimpCryptid — CHANGES_REQUESTED** [review]
Thanks for the revision! Please see the attached comments and resubmit.
Will test the gamePage.ts section once these initial fixes are made.

**Pull request closed without merging**